File: src/snmp/system.rs
use async_snmp::Client;

use crate::snmp::oid;

#[derive(Debug)]
pub struct SystemInfo {
    pub hostname: Option<String>,
    pub description: Option<String>,
    pub object_id: Option<String>,
}

pub async fn get_device_info(client: &Client) -> anyhow::Result<SystemInfo> {
    let mut sys_name = client.get(&oid::sys_name()).await?;
    let mut sys_descr = client.get(&oid::sys_descr()).await?;
    let mut sys_object_id = client.get(&oid::sys_object_id()).await?;

    let hostname = sys_name.varbinds.pop().map(|vb| vb.value.to_string());
    let description = sys_descr.varbinds.pop().map(|vb| vb.value.to_string());
    let object_id = sys_object_id.varbinds.pop().map(|vb| vb.value.to_string());

    Ok(SystemInfo {
        hostname,
        description,
        object_id,
    })
}

File: oxian_py/snmp/system.py
from __future__ import annotations

from dataclasses import dataclass

from . import oid
from .client import SnmpClient


@dataclass
class SystemInfo:
    """SNMP System MIB-II summary."""

    hostname: str | None = None
    description: str | None = None
    object_id: str | None = None


async def get_device_info(client: SnmpClient) -> SystemInfo:
    """Query system name, description, and sysObjectID from target device."""
    sys_name = await client.get(oid.sys_name())
    sys_descr = await client.get(oid.sys_descr())
    sys_object_id = await client.get(oid.sys_object_id())

    hostname = str(sys_name).strip() if sys_name is not None else None
    description = str(sys_descr).strip() if sys_descr is not None else None
    object_id = str(sys_object_id).strip() if sys_object_id is not None else None

    return SystemInfo(
        hostname=hostname,
        description=description,
        object_id=object_id,
    )
